papermill/hdfs.py

```diff
@@ -0,0 +1,157 @@
+        except FileNotFound:
+            return False
+
+    def _is_dir(self, path):
+        """:param path:str path which needs to be checked if it's a directory
+           :returns true if the given path is directory"""
+        try:
+            file_status = self.__webhdfs.get_file_dir_status(path)
+            return file_status['FileStatus']['type'] == 'DIRECTORY'
+        except FileNotFound:
+            return False
+
+    def _is_file(self, path):
+        """:param path:str path which needs to be checked if it's a directory
+           :returns true if the given path is file"""
+        try:
+            file_status = self.__webhdfs.get_file_dir_status(path)
+            return file_status['FileStatus']['type'] == 'FILE'
+        except FileNotFound:
+            return False
+
+    def _is_empty(self, path):
+        """:param path:str path which needs to be checked if it's a directory
+           :returns true if the given path is empty else false"""
+        try:
+            file_status = self.__webhdfs.get_file_dir_status(path)
+            if len(file_status) > 0:
+                return False
+            else:
+                return True
+        except Exception as e:
+            raise e
+
+    def _path_basename(self, path):
+        """:param path:str
+           :returns base path of the given path"""
+        base_path = path.split(self._path_separator)[:-1]
+        return '/'.join(base_path)
+
+    @retry(3)
+    def _list(self, path):
+        """:param path:str
+           :returns lists the files in the directory"""
+        if self._exists(path) and not self._is_file(path):
+            hdfs_files = self.__webhdfs.list_dir(path)
+            if not path.endswith('/'):
+                path += '/'
+            for f in hdfs_files['FileStatuses']['FileStatus']:
+                yield path + f['pathSuffix']
+        else:
+            raise FileNotFound("directory {directory} doesn't exist".format(directory=path))
+
+    def _read(self, path, size=-1, offset=0, binary=False):
+        """:param path:str
+           :returns reads contents of the given file path and returns"""
+        if self._exists(path) and not self._is_file(path):
+            raise IOError("file {path} doesn't exist".format(path=path))
+        read = self.__webhdfs.read_file(path, offset=offset, length='null' if size < 0 else size)
+        if not binary:
+            read = read.decode('utf-8')
+        return read
+
+    def _write(self, path, s=""):
+        """:param path:str
+           :returns writes the given content to file. By default it'll overwrite the existing file's
+        contents. Can be disabled."""
+        if not self._exists(path) and self._exists(self._path_basename(path)):
+            return self.__webhdfs.create_file(path, s, overwrite=True)
+        else:
+            raise FileNotFound("parent directory {} doesn't exist".format(self._path_basename(path)))
+
+    def _mkdir(self, path, recursive, exist_ok=False):
```

Comment:
You don't actually use `recursive` except to raise exceptions? I'd drop `exist_ok` and just assume it's True for execution purposes